create unit test guideline (#869)
vdaas/vald: docs/contributing/unit-test-guideline.md

# Unit Test Guideline

## About

> When we create the unit test for each function or package, we should consider test coverage not only code coverage.

Author: 🤔

> When we create the unit test for each function or package, we should consider test coverage not only code coverage.

Reviewer: @vankichi what do you think 🤔

> When we create the unit test for each function or package, we should consider test coverage not only code coverage.

Author: what is the definition of quality?

> When we create the unit test for each function or package, we should consider test coverage not only code coverage.

Author: Updated, please check the diff

> When we create the unit test for each function or package, we should consider test coverage not only code coverage.

Reviewer: sorry my mistake. 
`definition of quality` is not needed in unit test guideline

This guideline will help you to create good unit tests.

## Unit Test

Before considering the unit test, we should define what is the unit.
Unit is a set of procedures or functions, in a procedural or functional language.

> Also, it is a class and its needed classes, in an object or object-oriented language.

Author: e.g. extension class

> Also, it is a class and its needed classes, in an object or object-oriented language.

Reviewer: ```suggestion
Also, it is a class and its dependencies, in an object or object-oriented language.
```

> Also, it is a class and its needed classes, in an object or object-oriented language.

Author: dependencies has more meaning.
we should not use this word.

Unit testing validates the basic units of the program in isolation.

## Guideline

### Strategy

Before creating the unit tests, we must know the essentials of strategy.

1. Know what you are testing

> 	- Purpose/Consumers of target function or classes or etc / designed by contract  defensive program.

Reviewer 2: ```suggestion
	- Purpose/Consumers of target function or classes or etc / designed by contract defensive program.
```

1. Test behaviors and results, not implementation
	- ensures that tests only fail when there is an actual effect and not due to internal changes.
1. Test one thing at a time
	- Each test should have a clear, concise, singular objective.
1. Make tests readable and understandable
	- Hold test code to a similar standard as production code.
1. Make tests deterministic
	- A test should pass all the time or fail all the time until fixed.
1. Make tests independent and self-sufficient
	- Setup, execution, and verification steps in a given test should not depend on running other tests before it. To keep unit tests simple, fast running, and easy to debug, it may be necessary to isolate the class under test.
1. Repeat yourself when necessary
	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
1. Measure code coverage but focus on test coverage
	- Do not simply to archive code coverage.

### Test case

#### Basic

> It is not perfect, but we should try to cover all codes at first.

Author: How about it ?
```suggestion
It is not perfect to improve test coverage, but we should try to cover all codes at first.
```

> To cover all code, the basic way is the creating test cases to complete all branches in the target unit.

Reviewer: ```suggestion
To cover all code, the basic way is to create test cases to complete all branches in the target unit.
```

Let's see the below function.
```go
func calcSum(val ...int32) (sum int32) {
	if len(val) == 0 {
		return
	} else {
		for _, v := range val {
			sum += v
		}
		return
	}

> }

Reviewer 2: Thank you for your example.
First, I understood your opinion. And I agree with you.
But the checked example code, we can also represent like this more simpler. 

```go
func calcSum(val ...int32) (sum int32) {
	for _, v := range val {
		sum += v
	}
	return
}
```
In golang, Even if we pass nil or empty as a variadic argument, it is converted to an empty slice.
Therefore how about changing the `calcSum ` method to return an error like this?

```go
func calcSum(val ...int32) (sum int32, err error) {
	if len(val) == 0 {
		return 0, errors.New("empty")
	}
	for _, v := range val {
		sum += v
	}
	return sum, nil
}
```

Because it is necessary to check for errors and the writing style is unified, I think the following sentences will be conveyed more clearly.

59-64
> When the above function is given, we should create 2 test for achieving 100% code coverage at least.
> - When val is not given
>   - In other words, the default value is given
> - When val(int32) is given
>  - `var val int32 = 1`

If I made a mistake, I am sorry :bow:

Reviewer: @hlts2 I am not sure if returning error in this example is necessary or not. I think it depends on the program design.

Reviewer 2: @kevindiu Thank you. I got it 👍 

> it depends on the program design.

I think so too. Is there any other example? 🤔

Author: Thx ,but why did you think so.
I just thought this is just the example code for explaining input values of base strategy.
I'd like to remove any  codes which is not related this explanation.
could you please read this draft document over all again?
and if you suggest about example code, please keep in your mind it is needed only for describe basic input patterns.

Reviewer 2: > I just thought this is just the example code for explaining input values of base strategy.

I got it. thank you.

```

> When the above function is given, we should create 2 test for achieving 100% code coverage at least.

Reviewer: ```suggestion
When the above function is given, we should create 2 tests to achieving 100% code coverage at least.
```

- When val is not given
  - In other words, the default value is given
- When val(int32) is given
  - `var val int32 = 1`
  
It seems enough for the given function, but we have to take care of some test cases that are remaining for improving test coverage.

> In this case, there is one test case is remaining.

Reviewer: ```suggestion
In this case, there are at least one more test case to complete test coverage.
```

Author: ```suggestion
In this case, there is at least one more test case to complete test coverage.
```
@kevindiu i checked grammar

- When val([]int32) is given

>   - `var val []int32 = {1, 2}`

Reviewer: how about boundary values, overflow, and negative value?

Author: boundary values, overflow are OK, but negative values is not needed because of cannot be set.

Reviewer: I think it can be set because it is int32 not uint32 🤔
https://play.golang.org/p/3xfIIDU-D1G

Author: oh sorry, it's my mistake 😓 
I will add all patterns you commented.

  
That is the focus on test coverage.

> Therefore, we should consider all cases for improving test coverage of the target unit.

Reviewer: ```suggestion
Therefore, we should consider all cases to improve test coverage for the target unit.
```

To improve test coverage, the basic but critical thinking way is thinking about input patterns.
It is not only a single input, but also multi inputs.

> If a function or method requires multi-input, we should try to create many test cases.

Reviewer: ```suggestion
If a function or method accept multiple inputs, we should try to create test cases to cover all the inputs.
```

(Note: The below `calcSum()` is diffrent function from `calsSum(val ...int32)` which we mention before.)

```go
func calcAverageDiff(val1 []int32, val2 []int32) (diff float64) {
	var ave1, ave2 float64
	if len(val1) == 0 && len(val2) == 0 {
		return
	}
	if len(val1) != 0 {
		ave1 = float64(calcSum(val1)) / float64(len(val1))
	}
	if len(val2) != 0 {
		ave2 = float64(calcSum(val2)) / float64(len(val2))
	}

	diff = math.Abs(ave1 - ave2)
	return
}
```

When `calcAverageDiff()` is given, the test patterns are below:

| len(val1) | len(val2) | option |
|:-----:|:-----:|:-----:|
| 0  | 0  | - |
| 0  | >0 | - |
| >0 | 0  | - |
| >0 | >0 | ave1 > ave2 |

> | >0 | >0 | ave1 < ave2 |

Reviewer: sane as before, how about boundary values, overflow, and negative value?

Author: I think it is not needed.
Both `ave1` and `ave2` are calculated variable, so we do not have to test all of you mentioned.
At least, we have to test overflows.
btw, what did you mean boundary values?

Reviewer: boundary values means the Max/Min values. I think you have covered it already.

but do we need to cover the overflow cases here? as `calcSum` will overflow the value

Author: I could not reproduce such overflow matter.
could you show me in the playground?

Reviewer: for example this.
https://play.golang.org/p/mIiLeW6Qf9j
🤔

Author: thx. I got it.

At a glance, one of the last 2 patterns is enough, but these will help us to notice the bug in dependency.
It will avoid the unexpected error due to update dependencies.

> You should create unit tests for error patterns as same as successive patterns.

Reviewer: ```suggestion
You should create unit tests for error patterns as the same as success patterns.
```

Author: ```suggestion
You have to create unit tests for error patterns as the same as success patterns.
```

#### Advanced

##### Robust boudary test

> The previous section is about a basic test case.

Reviewer: ```suggestion
The previous section is about the basic test cases.
```

> For more cover test coverage, the (robust) boundary test should be applied.

Reviewer: ```suggestion
To cover more test coverage, the (robust) boundary test should be applied.
```

A boundary test is testing values on or near the boundaries of allowed inputs.

```go
func AgeTest(val int32) (ok bool) {
	if val >= 20 && val < 65 {
		ok = true
	}
	return
}
```

When the above function is given as target, the minimum number of tests based on the "Robust Boundary" test is 7.
  - the input `val` cases: 19, 20, 21, 30, 64, 65, 66
  
The robust boundary test requires 6N+1 test cases (N is the number of inputs for target function).

> More increasing the required input, more increasing the minimum number of cases.

Reviewer: ```suggestion
Increasing the input arguments will increase the number of test cases.
```

Author: ```suggestion
Increasing the input arguments will increase the number of needed test cases.
```

Reviewer: ```suggestion
Increasing the input arguments will also increase the number of required test cases.
```

Sometimes, it requires a lot of resources to create/maintain test.
So, you may test the critical cases using boundary value and have a sense of purpose.

#### Equivalence Class Testing

Equivalence Class Testing is grouping the values together such that all members of the group are considered the same and testing one value from each.
```go
func AgeTest(val int32) (ok bool) {
	if val >= 20 && val < 65 {
		ok = true
	}
	return
}
```

When the above function is given as target, you have to create 3 groups and pick one value from each group.

|Desc|Input range|Chose Input|
|:-----:|:-----:|:-----:|
| Under 20    | {0, ...,  19 }| 5  |
| worker range| {20, ..., 64} | 29 |
| Senior      | {65, ...    } | 72 |

### Vald Style

In the Vald, we create unit tests based on the basic test case.

> And, you also create unit tests based on robust boundary tests or equivalence class tests as needed.

Reviewer: ```suggestion
And, you may also create unit tests based on robust boundary tests or equivalence class tests if needed.
```

But, we have to take care about that the Vald is developed using Go.
As you know as Go has many coding features as other languages.

> One of the features is the Go will convert a single value to a slice value when the Function or Method receives variadic argument (e.g. `...[]int`, `...[]string`, `...interface{}`, or etc) as the input.

Reviewer: I think many languages support it. If talking about unique feature of go, it should be `channel`

Author: yes but it is one of the that.
I'd like to mention for below sentences.

> It is clear for Go coders. 

Reviewer: ```suggestion
```

And we apply the table-driven test for running unit tests.
For example, when we create the unit test of `func getMeta(...[]int)`, the test code will be more complex than other functions' test which don't use variadic argument as the input if we create the test for all input patterns.
Considering those, we define the basic unit case which is a little diffrent from [the basic test case](#Basic).

This change is very clear and you can apply it easily.
Our basic test case depends on the type of 2 variadic argument.

  1. When input is `...interface{}`

>       - we have to all test cases with value which satisfies `...interface{}` as same as [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.

Reviewer: ```suggestion
      - we have to write all test cases which satisfies `...interface{}` as same as [basic test case](#Basic). For example, `val = 1`, `val = "input"`, `val = []float64{2020.12}` and so on.
```

  1. When input is not `...interface{}` but `...[]int`, `...[]string` or etc
      - we have to create only slice pattern test cases, which is the same as not create test cases with a single vale.

>       - we should test with boundary cases, for example, we should test with `val = []int{math.MaxInt64()}` when the input value is `...[]int`.

Reviewer: not sure if i am correct but the difference here is testing `interface{}` vs other types

Author: 🤔 
this section mentions about variadic argument

> Summarize Vlad unit test guideline:

Reviewer: ```suggestion
Summarize Vald unit test guideline:
```

# Unit Test Guideline

## About

The goal of the unit test is to check whether the unit is implemented correctly in various cases.
A good unit test is to check various cases even which we do not expect as supposed usage.
When we create the unit test for each function or package, we have to consider test coverage not only code coverage.
This guideline will help you to create good unit tests.

## Unit Test

Before considering the unit test, we should define what is the unit.
Unit is a set of procedures or functions, in a procedural or functional language.
Also, it is a class and its needed classes, in an object or object-oriented language.
Unit testing validates the basic units of the program in isolation.

## Guideline

### Strategy

Before creating the unit tests, we must know the essentials of strategy.

1. Know what you are testing

> 	- Purpose/Consumers of target function or classes or etc / designed by contract defensive program.

Reviewer: ```suggestion
	- Purpose/Consumers of target (function, classes and etc.) / designed by contract defensive program.
```

  - the input `val` cases: 19, 20, 21, 30, 64, 65, 66
  
The robust boundary test requires 6N+1 test cases (N is the number of inputs for target function).
Increasing the input arguments will also increase the number of required test cases.
Sometimes, it requires a lot of resources to create/maintain test.
So, you may test the critical cases using boundary value and have a sense of purpose.

#### Equivalence Class Testing

Equivalence Class Testing is grouping the values together such that all members of the group are considered the same and testing one value from each.
```go
func AgeTest(val int32) (ok bool) {
	if val >= 20 && val < 65 {
		ok = true
	}
	return
}
```

When the above function is given as target, you have to create 3 groups and pick one value from each group.

|Desc|Input range|Chose Input|
|:-----:|:-----:|:-----:|
| Under 20    | {0, ...,  19 }| 5  |
| worker range| {20, ..., 64} | 29 |
| Senior      | {65, ...    } | 72 |

### Vald Style

In the Vald, we create unit tests based on the basic test case.
And, you may also create unit tests based on robust boundary tests or equivalence class tests if needed.

> But, we have to take care about that the Vald is developed using Go.

Reviewer 2: ```suggestion
But, we have to take care that the Vald is developed using Go.
```

Author: ```suggestion
But, we have to take care of the Vald is developed using Go.
```

Reviewer 2: thank you for your suggestion. 🙇
It looks good to me.
I think your suggestion is good.
